tests/events/network_client/main.c

```diff
@@ -0,0 +1,200 @@
+#include <limits.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <time.h>
+#include <unistd.h>
+
+#include <stdio.h>
+
+#include "events.h"
+#include "parsenum.h"
+#include "sock.h"
+#include "warnp.h"
+
+static int done;
+
+static int
+poke(struct sock_addr ** sas)
+{
+	int client_sock;
+
+	/* Attempt to connect, but don't complain about any failures. */
+	if ((client_sock = sock_connect(sas)) == -1)
+		goto err0;
+
+	/* Don't send anything; just close the connection. */
+	if (close(client_sock)) {
+		warnp("close");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Wait duration can be interrupted by signals. */
+static int
+wait_ms(size_t msec)
+{
+	struct timespec ts;
+
+	ts.tv_sec = msec / 1000;
+	ts.tv_nsec = (msec % 1000) * 1000000;
+	return (nanosleep(&ts, NULL));
+}
+
+static int
+repeated_poker(struct sock_addr ** sas)
+{
+
+	/* "Poke" the server multiple times. */
+	do {
+		/*
+		 * We expect this to work a few times, then stop working
+		 * because the server shut itself down.  Verification will be
+		 * done via the logfile, so don't warn about errors here.
+		 */
+		if (poke(sas))
+			goto serverdone;
+
+		/* Wait a bit. */
+		if (wait_ms(100)) {
+			warnp("nanosleep");
+			goto err0;
+		}
+	} while (!done);
+
+serverdone:
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+static int
+long_poke(struct sock_addr ** sas, pid_t server_pid)
+{
+
+	/* Poke the server once. */
+	if (poke(sas)) {
+		warn0("poke");
+		goto err0;
+	}
+
+	/* Wait for the server to start its poll(). */
+	wait_ms(100);
+
+	/* Send a shutdown signal to the server. */
+	if (kill(server_pid, SIGUSR1)) {
+		warnp("kill");
+		goto err0;
+	}
+
+	/* Wait for the server to process the signal. */
+	wait_ms(100);
+
+	/* This should fail because the server shut itself down. */
+	if (!poke(sas)) {
+		warn0("poke should have failed");
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Parse a pid_t; it's a signed integer and there's no PID_MAX. */
+static pid_t
+str_to_pid(const char * str)
```

Comment:
This is yucky, but it's the cleanest solution I could find for reading a pid from the command-line.

Yes, POSIX 2008 [1] includes
> The implementation shall support one or more programming environments in which the widths of blksize_t, pid_t, size_t, ssize_t, and suseconds_t are no greater than the width of type long.

but I don't think that means that we can assume that our current programming environment qualifies.  The paragraph goes on to say:

> The names of these programming environments can be obtained using the confstr() function or the getconf utility.

and we're not checking either of those.

[1] https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html